src/Data/List.purs

```diff
@@ -794,3 +845,62 @@ transpose ((x : xs) : xss) =
 foldM :: forall m a b. Monad m => (b -> a -> m b) -> b -> List a -> m b
 foldM _ b Nil = pure b
 foldM f b (a : as) = f b a >>= \b' -> foldM f b' as
+
+--------------------------------------------------------------------------------
+-- Fast operations which also reverse the list ---------------------------------
+--------------------------------------------------------------------------------
+
+-- | Maps a function to each element in a list
+-- | and reverses the result, but faster than
+-- | running each separately. Equivalent to:
+-- |
+-- | ```purescript
+-- | \f l = map f l # reverse
+-- | ```
+-- |
+-- | Running time: `O(n)`
+mapReverse :: forall a b. (a -> b) -> List a -> List b
+mapReverse f = go Nil
+  where
+  go :: List b -> List a -> List b
+  go acc Nil = acc
+  go acc (x : xs) = go (f x : acc) xs
+
+-- | Converts each element to a Tuple containing its index,
+-- | and reverses the result, but faster than running separately.
+-- | Equivalent to:
+-- |
+-- | ```purescript
+-- | reverse <<< mapWithIndex Tuple
+-- | ```
+-- |
+-- | Running time: `O(n)`
+addIndexReverse :: forall a. List a -> List (Tuple Int a)
+addIndexReverse = go 0 Nil
+  where
+  go :: Int -> List (Tuple Int a) -> List a -> List (Tuple Int a)
+  go i acc Nil = acc
+  go i acc (x : xs) = go (i + 1) ((Tuple i x) : acc) xs
+
+-- | Removes neighboring duplicate items from a list
+-- | based on an equality predicate.
+-- | Keeps the LAST element if duplicates are encountered.
+-- | Returned list is reversed (this is to improve performance).
+-- |
+-- | ```purescript
+-- | nubByAdjacentReverse (on eq length) ([1]:[2]:[3,4]:Nil) == [3,4]:[2]:Nil`
+-- | ```
+-- |
+-- | Running time: `O(n)`
+nubByAdjacentReverse :: forall a. (a -> a -> Boolean) -> List a -> List a
+nubByAdjacentReverse p = go Nil
+  where
+    go :: List a -> List a -> List a
+    -- empty output
+    go Nil (x : xs) = go (x : Nil) xs
+    -- checking for duplicates
+    go acc@(a : as) (x : xs)
+      | p a x = go (x : as) xs
+      | otherwise = go (x : acc) xs
+    -- empty input
+    go acc Nil = acc
```

Comment:
Elsewhere the `where` definitions aren't indented

```suggestion
  go :: List a -> List a -> List a
  -- empty output
  go Nil (x : xs) = go (x : Nil) xs
  -- checking for duplicates
  go acc@(a : as) (x : xs)
    | p a x = go (x : as) xs
    | otherwise = go (x : acc) xs
  -- empty input
  go acc Nil = acc
```